src/nix/flake.cc

```diff
@@ -334,10 +334,10 @@ struct CmdFlakeCheck : FlakeCommand
             try {
                 state->forceValue(v, pos);
                 if (!v.isLambda() || v.lambda.fun->matchAttrs || std::string(v.lambda.fun->arg) != "final")
-                    throw Error("overlay does not take an argument named 'final'");
+                    throw Error("By convention the attribute 'overlay' expects the argument 'final' not '" + std::string(v.lambda.fun->arg) + "'");
```

Comment:
This message isn't right for "matchAttrs" type lambdas, such as `{ foo, bar }: x`. That case should have its own message.
Same below.

Reply:
@roberth Do you have a proposal for what error it should throw?

Reply:
By convention the attribute 'overlay' is expected to be a simple lambda starting with 'final: ', not an attribute set matching lambda.

I guess you could add a representation of it too, but a source position `Pos` would be even better.

Reply:
Position info is already added to the error context by the `try { ... } catch { ... }`.